music_assistant/providers/radioparadise/__init__.py

```diff
@@ -0,0 +1,513 @@
+            song_start = int(song.get("elapsed", 0))
+            song_duration = int(song.get("duration", 0))
+            song_end = song_start + song_duration
+
+            if song_start <= current_time_ms < song_end:
+                return song
+
+        # If no exact match, return first song
+        first_song = songs.get("0")
+        return cast("dict[str, Any]", first_song) if first_song is not None else None
+
+    def _get_next_song(
+        self, songs: dict[str, Any], current_song: dict[str, Any]
+    ) -> dict[str, Any] | None:
+        """Get the next song after current song."""
+        current_event = current_song.get("event")
+        for song_key in sorted(songs.keys(), key=int):
+            song = cast("dict[str, Any]", songs[song_key])
+            if song.get("event") != current_event and int(song.get("elapsed", 0)) > int(
+                current_song.get("elapsed", 0)
+            ):
+                return song
+        return None
+
+    def _build_stream_url(self, channel_id: str) -> str:
+        """Build stream URL for a channel with fallback to other formats."""
+        if channel_id not in self._channels_cache:
+            return ""
+
+        channel_info = self._channels_cache[channel_id]
+        stream_urls = channel_info.get("stream_urls", {})
+
+        try:
+            current_format_index = FALLBACK_ORDER.index(self._stream_format)
+        except ValueError:
+            # If preferred format not in fallback order, start from beginning
+            current_format_index = 0
+
+        for format_key in FALLBACK_ORDER[current_format_index:]:
+            if format_key in stream_urls:
+                return cast("str", stream_urls[format_key])
+
+        return ""
+
+    async def _monitor_stream_metadata(self, stream_details: StreamDetails) -> None:
+        """Monitor and update the StreamDetails with rich metadata every 10 seconds."""
+        last_track_event = ""
+        item_id = stream_details.item_id
+
+        try:
+            while True:
+                metadata = await self._get_channel_metadata(item_id)
+                if metadata and metadata.get("current"):
+                    current_song = metadata["current"]
+                    current_event = current_song.get("event", "")
+
+                    if current_event != last_track_event:
+                        # Create StreamMetadata object with full track info
+                        stream_metadata = self._build_stream_metadata(current_song, metadata)
+
+                        self.logger.info(
+                            f"Updating stream metadata for {item_id}: "
+                            f"{stream_metadata.artist} - {stream_metadata.title}"
+                        )
+                        stream_details.stream_metadata = stream_metadata
+
+                        last_track_event = current_event
+
+                await asyncio.sleep(10)
+        except Exception:
```

Comment:
if possible, try to avoid these broad exception handling and instead use a more narrow exception you'd like to catch. For instance, using aiohttp.ClientError catches any issues that arise during fetching the data from the web